### ISSUE TITLE
Tensors Sizes sometimes on Eval

### ISSUE THREAD

**Issue by lockder** [OpenNMT/OpenNMT-tf#142]
I am doing character translation and when i am training everything is fine. But when i do the eval i found some times while using the beam search i get a tensor size error. Tensorflow say they are expecting some size but they got another size. Its wierd because its not always but very often. While training everything is fine. Any suggestions?

**Comment by guillaumekln** [comment]
Can you copy-paste the error message?

**Comment by lockder** [comment]
Sure I will send you all the information.
if I set the beam_width to 1 it happens much less.

I added the data I'm using.
I'm translating human time format to machine time format.
My language is spanish but its easy to see the data :)
[data.zip](https://github.com/OpenNMT/OpenNMT-tf/files/2061820/data.zip)


##### The Model:

class CharacterSeq2Seq(onmt.models.SequenceToSequence):
  """Defines a small unidirectional LSTM encoder-decoder model."""
  def __init__(self):
    super(CharacterSeq2Seq, self).__init__(
        source_inputter=onmt.inputters.WordEmbedder(
            vocabulary_file_key="source_chars_vocabulary",
            embedding_size=100,
            tokenizer=CharacterTokenizer()),
        target_inputter=onmt.inputters.WordEmbedder(
            vocabulary_file_key="target_chars_vocabulary",
            embedding_size=100,
            tokenizer=CharacterTokenizer()),
        encoder=onmt.encoders.GoogleRNNEncoder(
            num_layers=2,
            num_units=512,
            dropout=0.3),
        decoder=onmt.decoders.AttentionalRNNDecoder(
            num_layers=2,
            num_units=512,
            bridge=onmt.layers.CopyBridge(),
            attention_mechanism_class=tf.contrib.seq2seq.BahdanauAttention,
            cell_class=tf.contrib.rnn.LSTMCell,
            dropout=0.2,
            residual_connections=False))


def model():
  return CharacterSeq2Seq()

####### The configuration

#Model Data
model_dir: model

data:
  train_features_file: source/raw.txt
  train_labels_file: target/raw.txt
  eval_features_file: source/test.txt
  eval_labels_file: target/test.txt
  source_chars_vocabulary: source/character/vocab.txt
  target_chars_vocabulary: target/character/vocab.txt

# Model and optimization parameters.
params:
  # The optimizer class name in tf.train or tf.contrib.opt.
  optimizer: AdamOptimizer
  # (optional) Additional optimizer parameters as defined in their documentation.
  #optimizer_params:
   # beta1: 0.8
    #beta2: 0.998
  learning_rate: 0.005

  # (optional) Global parameter initialization [-param_init, param_init].
  #param_init: 0.1
  # (optional) Maximum gradients norm (default: None).
  clip_gradients: 5.0
  # (optional) Average loss in the time dimension in addition to the batch dimension (default: False).
  average_loss_in_time: false
  # (optional) The type of learning rate decay (default: None). See:
  #  * https://www.tensorflow.org/versions/master/api_guides/python/train#Decaying_the_learning_rate
  #  * opennmt/utils/decay.py
  # This value may change the semantics of other decay options. See the documentation or the code.
  decay_type: exponential_decay
  # (optional unless decay_type is set) The learning rate decay rate.
  decay_rate: 0.7
  # (optional unless decay_type is set) Decay every this many steps.
  decay_steps: 100
  # (optional) The number of training steps that make 1 decay step (default: 1).
  decay_step_duration: 1
  # (optional) If true, the learning rate is decayed in a staircase fashion (default: True).
  staircase: true
  # (optional) After how many steps to start the decay (default: 0).
  start_decay_steps: 100
  # (optional) Stop decay when this learning rate value is reached (default: 0).
  minimum_learning_rate: 0.00001
  # (optional) Type of scheduled sampling (can be "constant", "linear", "exponential",
  # or "inverse_sigmoid", default: "constant").
  #scheduled_sampling_type: constant
  # (optional) Probability to read directly from the inputs instead of sampling categorically
  # from the output ids (default: 1).
  #scheduled_sampling_read_probability: 1
  # (optional unless scheduled_sampling_type is set) The constant k of the schedule.
  #scheduled_sampling_k: 0
  # (optional) The label smoothing value.
  #label_smoothing: 0.1
  # (optional) Width of the beam search (default: 1).
  beam_width: 5
  # (optional) Length penaly weight to apply on hypotheses (default: 0).
  #length_penalty: 0.2
  # (optional) Maximum decoding iterations before stopping (default: 250).
  maximum_iterations: 250

# Training options.
train:
  batch_size: 128

  # (optional) Batch size is the number of "examples" or "tokens" (default: "examples").
  batch_type: examples
  # (optional) Save a checkpoint every this many steps.
  save_checkpoints_steps: 500
  # (optional) How many checkpoints to keep on disk.
  keep_checkpoint_max: 1
  # (optional) Save summaries every this many steps.
  save_summary_steps: 100
  # (optional) Train for this many steps. If not set, train forever.
  train_steps: 1500
  # (optional) The width of the length buckets to select batch candidates from (default: 5).
  bucket_width: 1
  # (optional) The number of threads to use for processing data in parallel (default: 4).
  num_threads: 1
  # (optional) The number of elements from which to sample during shuffling (default: 500000).
  # Consider setting this to the number of training examples.
  sample_buffer_size: 50000
  # (optional) The number of batches to prefetch asynchronously (default: 1).
  prefetch_buffer_size: 1

# (optional) Evaluation options.
eval:
  # (optional) The batch size to use (default: 32).
  batch_size: 128
  # (optional) The number of threads to use for processing data in parallel (default: 1).
  num_threads: 8
  # (optional) The number of batches to prefetch asynchronously (default: 1).
  prefetch_buffer_size: 1
  # (optional) Evaluate every this many seconds (default: 18000).
  eval_delay: 3000
  # (optional) Save evaluation predictions in model_dir/eval/.
  save_eval_predictions: false
  # (optional) Evalutator or list of evaluators that are called on the saved evaluation predictions.
  # Available evaluators: BLEU, BLEU-detok
  external_evaluators: BLEU

# (optional) Inference options.
infer:
  # (optional) The batch size to use (default: 1).
  batch_size: 1
  # (optional) The number of threads to use for processing data in parallel (default: 1).
  num_threads: 1
  # (optional) The number of batches to prefetch asynchronously (default: 1).
  prefetch_buffer_size: 1
  # (optional) For compatible models, the number of hypotheses to output (default: 1).
  n_best: 1


#### The crash

INFO:tensorflow:Loading serialized model description from model/model_description.pkl
2018-06-01 10:27:57.278068: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
INFO:tensorflow:Using config: {'_model_dir': 'model', '_tf_random_seed': None, '_save_summary_steps': 100, '_save_checkpoints_steps': 500, '_save_checkpoints_secs': None, '_session_config': gpu_options {
}
allow_soft_placement: true
, '_keep_checkpoint_max': 1, '_keep_checkpoint_every_n_hours': 10000, '_log_step_count_steps': 100, '_train_distribute': None, '_service': None, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x11f6868d0>, '_task_type': 'worker', '_task_id': 0, '_global_id_in_cluster': 0, '_master': '', '_evaluation_master': '', '_is_chief': True, '_num_ps_replicas': 0, '_num_worker_replicas': 1}
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2018-06-01-08:27:58
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from model/model.ckpt-1500
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
2018-06-01 10:28:06.245988: W tensorflow/core/framework/op_kernel.cc:1318] OP_REQUIRES failed at gather_nd_op.cc:50 : Invalid argument: flat indices[9759, :] = [15, 31, 6] does not index into param (shape: [81,128,5,41]).
Traceback (most recent call last):
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1322, in _do_call
    return fn(*args)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1307, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1409, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.InvalidArgumentError: flat indices[9759, :] = [15, 31, 6] does not index into param (shape: [81,128,5,41]).
	 [[Node: seq2seq/decoder_1/decoder_1/GatherNd = GatherNd[Tindices=DT_INT32, Tparams=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"](seq2seq/decoder_1/decoder_1/Reshape_2, seq2seq/decoder_1/decoder_1/stack)]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/time_Translation/evaluate.py", line 24, in <module>
    main()
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/time_Translation/evaluate.py", line 20, in main
    evaluate()
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/time_Translation/evaluate.py", line 17, in evaluate
    nnmtMain()
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/opennmt/bin/main.py", line 125, in main
    runner.evaluate(checkpoint_path=args.checkpoint_path)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/opennmt/runner.py", line 157, in evaluate
    eval_spec.input_fn, hooks=eval_spec.hooks, checkpoint_path=checkpoint_path)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 425, in evaluate
    name=name)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 1117, in _evaluate_model
    config=self._session_config)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/training/evaluation.py", line 212, in _evaluate_once
    session.run(eval_ops, feed_dict)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 567, in run
    run_metadata=run_metadata)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1043, in run
    run_metadata=run_metadata)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1134, in run
    raise six.reraise(*original_exc_info)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1119, in run
    return self._sess.run(*args, **kwargs)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1191, in run
    run_metadata=run_metadata)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 971, in run
    return self._sess.run(*args, **kwargs)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 900, in run
    run_metadata_ptr)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1135, in _run
    feed_dict_tensor, options, run_metadata)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1316, in _do_run
    run_metadata)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1335, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InvalidArgumentError: flat indices[9759, :] = [15, 31, 6] does not index into param (shape: [81,128,5,41]).
	 [[Node: seq2seq/decoder_1/decoder_1/GatherNd = GatherNd[Tindices=DT_INT32, Tparams=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"](seq2seq/decoder_1/decoder_1/Reshape_2, seq2seq/decoder_1/decoder_1/stack)]]

Caused by op 'seq2seq/decoder_1/decoder_1/GatherNd', defined at:
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/time_Translation/evaluate.py", line 24, in <module>
    main()
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/time_Translation/evaluate.py", line 20, in main
    evaluate()
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/time_Translation/evaluate.py", line 17, in evaluate
    nnmtMain()
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/opennmt/bin/main.py", line 125, in main
    runner.evaluate(checkpoint_path=args.checkpoint_path)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/opennmt/runner.py", line 157, in evaluate
    eval_spec.input_fn, hooks=eval_spec.hooks, checkpoint_path=checkpoint_path)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 425, in evaluate
    name=name)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 1087, in _evaluate_model
    features, labels, model_fn_lib.ModeKeys.EVAL, self.config)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 831, in _call_model_fn
    model_fn_results = self._model_fn(features=features, **kwargs)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/opennmt/models/model.py", line 113, in _model_fn
    logits, predictions = self._build(features, labels, params, mode, config=config)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/opennmt/models/sequence_to_sequence.py", line 185, in _build
    return_alignment_history=True))
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/opennmt/decoders/rnn_decoder.py", line 254, in dynamic_decode_and_search
    decoder, maximum_iterations=maximum_iterations)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/contrib/seq2seq/python/ops/decoder.py", line 314, in dynamic_decode
    final_outputs, final_state, final_sequence_lengths)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py", line 458, in finalize
    final_state.cell_state))
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/util/nest.py", line 375, in map_structure
    structure[0], [func(*x) for x in entries])
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/util/nest.py", line 375, in <listcomp>
    structure[0], [func(*x) for x in entries])
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py", line 457, in <lambda>
    t, outputs.parent_ids, final_state.lengths),
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py", line 620, in _maybe_sort_array_beams
    return gather_tree_from_array(t, parent_ids, sequence_length)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py", line 180, in gather_tree_from_array
    ordered = array_ops.gather_nd(gather_from, indices)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/ops/gen_array_ops.py", line 2975, in gather_nd
    "GatherNd", params=params, indices=indices, name=name)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3392, in create_op
    op_def=op_def)
  File "/Users/sergio/cognitive/NaturalLanguageRecognition/NaturalLanguageManager/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1718, in __init__
    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

InvalidArgumentError (see above for traceback): flat indices[9759, :] = [15, 31, 6] does not index into param (shape: [81,128,5,41]).
	 [[Node: seq2seq/decoder_1/decoder_1/GatherNd = GatherNd[Tindices=DT_INT32, Tparams=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"](seq2seq/decoder_1/decoder_1/Reshape_2, seq2seq/decoder_1/decoder_1/stack)]]


Process finished with exit code 1

**Comment by lordzuko** [comment]
I am also having some similar issue, I trained my model on `tensorflow-gpu==1.4.0`, while running inferring I was getting following:
```
(translation) himanshu@StaquResearch2:~/virtualenv/translation/translation/OpenNMT-tf$ onmt-main infer --config config/opennmt-defaults.yml data/mandarin-english/translation.yml --features_file data/mandarin-english/evaluation.txt 
INFO:tensorflow:Loading serialized model description from data/mandarin-english-model/model_description.pkl
2018-06-04 10:46:29.534325: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
INFO:tensorflow:Using config: {'_service': None, '_keep_checkpoint_max': 5, '_save_checkpoints_steps': 5000, '_session_config': gpu_options {
}
allow_soft_placement: true
, '_master': '', '_save_checkpoints_secs': None, '_log_step_count_steps': 50, '_is_chief': True, '_model_dir': 'data/mandarin-english-model', '_task_id': 0, '_save_summary_steps': 50, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x7f947d75def0>, '_tf_random_seed': None, '_keep_checkpoint_every_n_hours': 10000, '_task_type': 'worker', '_num_ps_replicas': 0, '_num_worker_replicas': 1}
WARNING:tensorflow:The current version of tf.contrib.seq2seq.BeamSearchDecoder does not support returning the alignment history. None will be returned instead. Consider upgrading TensorFlow.
INFO:tensorflow:Restoring parameters from data/mandarin-english-model/model.ckpt-1000000
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>

```
As it suggested that BeamSearch Decoder was not found so I upgraded the tensorflow to `tensorflow==1.8.0` , but now I am getting similar error :

```
(translation) himanshu@StaquResearch2:~/virtualenv/translation/translation/OpenNMT-tf$ onmt-main infer --config config/opennmt-defaults.yml data/mandarin-english/translation.yml --features_file data/mandarin-english/evaluation.txt 
INFO:tensorflow:Loading serialized model description from data/mandarin-english-model/model_description.pkl
2018-06-04 10:52:17.416148: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
INFO:tensorflow:Using config: {'_keep_checkpoint_every_n_hours': 10000, '_is_chief': True, '_num_worker_replicas': 1, '_session_config': gpu_options {
}
allow_soft_placement: true
, '_master': '', '_save_checkpoints_steps': 5000, '_save_summary_steps': 50, '_model_dir': 'data/mandarin-english-model', '_task_type': 'worker', '_train_distribute': None, '_save_checkpoints_secs': None, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x7f09de4f84a8>, '_tf_random_seed': None, '_task_id': 0, '_keep_checkpoint_max': 5, '_service': None, '_global_id_in_cluster': 0, '_log_step_count_steps': 50, '_num_ps_replicas': 0, '_evaluation_master': ''}
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from data/mandarin-english-model/model.ckpt-1000000
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
2018-06-04 10:52:21.658949: W tensorflow/core/framework/op_kernel.cc:1318] OP_REQUIRES failed at gather_nd_op.cc:50 : Invalid argument: flat indices[195, :] = [3, 9, 6] does not index into param (shape: [7,10,5,1]).
Traceback (most recent call last):
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/client/session.py", line 1322, in _do_call
    return fn(*args)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/client/session.py", line 1307, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/client/session.py", line 1409, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.InvalidArgumentError: flat indices[195, :] = [3, 9, 6] does not index into param (shape: [7,10,5,1]).
	 [[Node: seq2seq/decoder_1/decoder_1/GatherNd = GatherNd[Tindices=DT_INT32, Tparams=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"](seq2seq/decoder_1/decoder_1/Reshape_2, seq2seq/decoder_1/decoder_1/stack)]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/himanshu/virtualenv/translation/bin/onmt-main", line 11, in <module>
    load_entry_point('OpenNMT-tf==1.4.1', 'console_scripts', 'onmt-main')()
  File "/home/himanshu/virtualenv/translation/translation/OpenNMT-tf/opennmt/bin/main.py", line 137, in main
    log_time=args.log_prediction_time)
  File "/home/himanshu/virtualenv/translation/translation/OpenNMT-tf/opennmt/runner.py", line 216, in infer
    hooks=infer_hooks):
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/estimator/estimator.py", line 509, in predict
    preds_evaluated = mon_sess.run(predictions)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/training/monitored_session.py", line 567, in run
    run_metadata=run_metadata)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/training/monitored_session.py", line 1043, in run
    run_metadata=run_metadata)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/training/monitored_session.py", line 1134, in run
    raise six.reraise(*original_exc_info)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/six.py", line 693, in reraise
    raise value
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/training/monitored_session.py", line 1119, in run
    return self._sess.run(*args, **kwargs)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/training/monitored_session.py", line 1191, in run
    run_metadata=run_metadata)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/training/monitored_session.py", line 971, in run
    return self._sess.run(*args, **kwargs)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/client/session.py", line 900, in run
    run_metadata_ptr)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/client/session.py", line 1135, in _run
    feed_dict_tensor, options, run_metadata)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/client/session.py", line 1316, in _do_run
    run_metadata)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/client/session.py", line 1335, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InvalidArgumentError: flat indices[195, :] = [3, 9, 6] does not index into param (shape: [7,10,5,1]).
	 [[Node: seq2seq/decoder_1/decoder_1/GatherNd = GatherNd[Tindices=DT_INT32, Tparams=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"](seq2seq/decoder_1/decoder_1/Reshape_2, seq2seq/decoder_1/decoder_1/stack)]]

Caused by op 'seq2seq/decoder_1/decoder_1/GatherNd', defined at:
  File "/home/himanshu/virtualenv/translation/bin/onmt-main", line 11, in <module>
    load_entry_point('OpenNMT-tf==1.4.1', 'console_scripts', 'onmt-main')()
  File "/home/himanshu/virtualenv/translation/translation/OpenNMT-tf/opennmt/bin/main.py", line 137, in main
    log_time=args.log_prediction_time)
  File "/home/himanshu/virtualenv/translation/translation/OpenNMT-tf/opennmt/runner.py", line 216, in infer
    hooks=infer_hooks):
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/estimator/estimator.py", line 496, in predict
    features, None, model_fn_lib.ModeKeys.PREDICT, self.config)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/estimator/estimator.py", line 831, in _call_model_fn
    model_fn_results = self._model_fn(features=features, **kwargs)
  File "/home/himanshu/virtualenv/translation/translation/OpenNMT-tf/opennmt/models/model.py", line 128, in _model_fn
    _, predictions = self._build(features, labels, params, mode, config=config)
  File "/home/himanshu/virtualenv/translation/translation/OpenNMT-tf/opennmt/models/sequence_to_sequence.py", line 185, in _build
    return_alignment_history=True))
  File "/home/himanshu/virtualenv/translation/translation/OpenNMT-tf/opennmt/decoders/rnn_decoder.py", line 258, in dynamic_decode_and_search
    decoder, maximum_iterations=maximum_iterations)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/contrib/seq2seq/python/ops/decoder.py", line 314, in dynamic_decode
    final_outputs, final_state, final_sequence_lengths)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py", line 458, in finalize
    final_state.cell_state))
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/util/nest.py", line 375, in map_structure
    structure[0], [func(*x) for x in entries])
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/util/nest.py", line 375, in <listcomp>
    structure[0], [func(*x) for x in entries])
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py", line 457, in <lambda>
    t, outputs.parent_ids, final_state.lengths),
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py", line 620, in _maybe_sort_array_beams
    return gather_tree_from_array(t, parent_ids, sequence_length)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py", line 180, in gather_tree_from_array
    ordered = array_ops.gather_nd(gather_from, indices)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/ops/gen_array_ops.py", line 2975, in gather_nd
    "GatherNd", params=params, indices=indices, name=name)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 3392, in create_op
    op_def=op_def)
  File "/home/himanshu/virtualenv/translation/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 1718, in __init__
    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

InvalidArgumentError (see above for traceback): flat indices[195, :] = [3, 9, 6] does not index into param (shape: [7,10,5,1]).
	 [[Node: seq2seq/decoder_1/decoder_1/GatherNd = GatherNd[Tindices=DT_INT32, Tparams=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"](seq2seq/decoder_1/decoder_1/Reshape_2, seq2seq/decoder_1/decoder_1/stack)]]

```

Can anyone tell me why this is happening ?

**Comment by guillaumekln** [comment]
Thanks for the details. The issue seems to happen when reordering the alignment history coming from the attention layer based on the beam search path. I will try to pinpoint to issue and submit a patch. This looks like a bug within TensorFlow.

In the meantime, I suggest manually disabling the alignment history so that this code path is not used:

```diff
diff --git a/opennmt/models/sequence_to_sequence.py b/opennmt/models/sequence_to_sequence.py
index f961e91..fc87900 100644
--- a/opennmt/models/sequence_to_sequence.py
+++ b/opennmt/models/sequence_to_sequence.py
@@ -182,7 +182,7 @@ class SequenceToSequence(Model):
                   memory=encoder_outputs,
                   memory_sequence_length=encoder_sequence_length,
                   dtype=target_dtype,
-                  return_alignment_history=True))
+                  return_alignment_history=False))
 
       target_vocab_rev = tf.contrib.lookup.index_to_string_table_from_file(
           self.target_inputter.vocabulary_file,
```

Let me know if this workaround works for you.

**Comment by guillaumekln** [comment]
Oh right. So let's revert this change and disable the flag at a lower level:

```diff
diff --git a/opennmt/decoders/rnn_decoder.py b/opennmt/decoders/rnn_decoder.py
index 9f7b229..7c7833f 100644
--- a/opennmt/decoders/rnn_decoder.py
+++ b/opennmt/decoders/rnn_decoder.py
@@ -217,7 +217,7 @@ class RNNDecoder(Decoder):
                       "be returned instead. Consider upgrading TensorFlow.")
       alignment_history = False
     else:
-      alignment_history = return_alignment_history
+      alignment_history = False
 
     batch_size = tf.shape(start_tokens)[0]
 ```

**Comment by lordzuko** [comment]
Now it's running, but it's printing nothing except <unk>, which means basically the output is all `None`.

```
(translation) himanshu@StaquResearch2:~/virtualenv/translation/translation/OpenNMT-tf$ onmt-main infer --config config/opennmt-defaults.yml data/mandarin-english/translation.yml --features_file data/mandarin-english/evaluation.txt 
INFO:tensorflow:Loading serialized model description from data/mandarin-english-model/model_description.pkl
2018-06-04 14:19:21.753868: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
INFO:tensorflow:Using config: {'_num_worker_replicas': 1, '_task_type': 'worker', '_task_id': 0, '_log_step_count_steps': 50, '_keep_checkpoint_every_n_hours': 10000, '_keep_checkpoint_max': 5, '_session_config': gpu_options {
}
allow_soft_placement: true
, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x7f2902cf82e8>, '_master': '', '_num_ps_replicas': 0, '_save_summary_steps': 50, '_evaluation_master': '', '_tf_random_seed': None, '_global_id_in_cluster': 0, '_model_dir': 'data/mandarin-english-model', '_save_checkpoints_steps': 5000, '_service': None, '_is_chief': True, '_train_distribute': None, '_save_checkpoints_secs': None}
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from data/mandarin-english-model/model.ckpt-1000000
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>
<unk> <unk> <unk> </s>

```

**Comment by guillaumekln** [comment]
I think this is an issue with the model you trained but it is unrelated to the error reported here. Maybe you want to open a separate issue, sharing your training parameters, the last training and eval loss, etc.

**Comment by lordzuko** [comment]
@guillaumekln When `return_alignment_history` is `False` is the beam search decoder being used for decoding or that makes it similar to the case when `beam_width = 1`

**Comment by guillaumekln** [comment]
The beam search decoder is used but will not return the alignment history produced by the attention layer.

**Comment by lockder** [comment]
Changing alignment_history = False, make it works :)

**Comment by lordzuko** [comment]
@lockder How is your translation result ?

**Comment by lockder** [comment]
@lordzuko Its working quite well. for now I'm just testing, it works better translating form human to machine language for time than for date. 
Since I'm doing character translation bleu will not work if I don't tokenize the eval files to setup every character like a word, so for now i'm using word_accuracy.
Time is almost 99.7 percent right now
Date is around 80 percent right now but I didn't tried to improve it yet

**Comment by R1j1t** [comment]
@guillaumekln I updated the bool mentioned above at 
```
diff --git a/opennmt/decoders/rnn_decoder.py b/opennmt/decoders/rnn_decoder.py
index 9f7b229..7c7833f 100644
--- a/opennmt/decoders/rnn_decoder.py
+++ b/opennmt/decoders/rnn_decoder.py
@@ -217,7 +217,7 @@ class RNNDecoder(Decoder):
                       "be returned instead. Consider upgrading TensorFlow.")
       alignment_history = False
     else:
-      alignment_history = return_alignment_history
+      alignment_history = False
 
     batch_size = tf.shape(start_tokens)[0]
```

But I am still getting the following error:

```
InvalidArgumentError (see above for traceback): flat indices[639, :] = [18, 1, 6
] does not index into param (shape: [19,7,5,15]).
         [[Node: seq2seq/decoder_1/decoder_1/GatherNd = GatherNd[Tindices=DT_INT
32, Tparams=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"](se
q2seq/decoder_1/decoder_1/Reshape_2, seq2seq/decoder_1/decoder_1/stack)]]
```
Could you please help.

**Comment by guillaumekln** [comment]
If someone is able and willing to share a checkpoint and input text for which this error is raised, that would be very helpful in trying to solve it. Thanks.

**Comment by Lawhy** [comment]
The checkpoint is too large (even after zip), so how to send you that?

**Comment by Lawhy** [comment]
The file is here:
https://drive.google.com/file/d/1JkuhyJYjRZ7oLnwrfDLPHzvJFEpRpnQd/view?usp=sharing

By the way, may I ask how to enable word features using openNMT-py, there is an instruction in Lua documentation but not in the pytorch one? Because I am labeling my characters with some annotation (both src and tgt side), I think this might help.

**Comment by lockder** [comment]
What I found is the bug happens to me only if I'm using tensorflow-cpu when I use tensorflow-gpu never crashes

**Comment by Lawhy** [comment]
I did use the cpu version. Thanks for your response. I have not tried it on gpu yet.

**Comment by guillaumekln** [comment]
Thank you for sharing the model, it was very helpful to debug.

I submitted a fix to TensorFlow: https://github.com/tensorflow/tensorflow/pull/20665

**Comment by guillaumekln** [comment]
The fix has been merged on the TensorFlow repository. It should be available in the next nightly package and, later, in the next TensorFlow major version.

Closing.
